sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala

```diff
@@ -236,15 +241,22 @@ private[sql] object PartitioningUtils {
   }
 
   /**
-   * Converts a string to a [[Literal]] with automatic type inference.  Currently only supports
-   * [[IntegerType]], [[LongType]], [[DoubleType]], [[DecimalType.SYSTEM_DEFAULT]], and
-   * [[StringType]].
+   * Converts a string to a [[Literal]] with automatic type inference if no field type specified.
+   * Auto inference only supports [[IntegerType]], [[LongType]], [[DoubleType]],
+   * [[DecimalType.SYSTEM_DEFAULT]], and [[StringType]].
    */
   private[sql] def inferPartitionColumnValue(
+      expectedDT: Option[DataType],
```

Comment:
I agree that casting to a non string type and then converting back to a string may lose precision, but what about disabling inference when calling inferPartitionColumnValue if the user has provided a schema? In that case, it should end up just returning the string literals, which you can then cast without a loss of precision.

Sent from my phone

> On Oct 20, 2015, at 8:00 AM, Cheng Hao <notifications@github.com> wrote:
> 
> In sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:
> 
> >     */
> >    private[sql] def inferPartitionColumnValue(
> > +      expectedDT: Option[DataType],
> We need to pass the expect the data type down and then get the associated literal-based partition column value; and @liancheng's suggestion kind of like get the literal (maybe string based) first, and then do casting outside, however, this probably lose some data precision during the re-casting.
> 
> For example:
> The path looks like, /part1=1.000, and with the auto inference, we will get a Double, and it will be cast to string as 1.0 if what user expect is StringType;
> 
> However, this is totally different if we get it as StringType directly, which supposed to be 1.000.
> 
> —
> Reply to this email directly or view it on GitHub.
>